test/jit/test_exception.py

```diff
@@ -0,0 +1,166 @@
+                    raise ArbitraryError
+            return a
+
+        with self.assertRaisesRegex(RuntimeError, "undefined value ArbitraryError"):
+            torch.jit.script(foo)
+
+        def exception_as_value():
+            a = Exception()
+            print(a)
+
+        with self.assertRaisesRegex(RuntimeError, "cannot be used as a value"):
+            torch.jit.script(exception_as_value)
+
+        @torch.jit.script
+        def foo_no_decl_always_throws():
+            raise RuntimeError("Hi")
+
+        # function that has no declared type but always throws set to None
+        output_type = next(foo_no_decl_always_throws.graph.outputs()).type()
+        self.assertTrue(str(output_type) == "NoneType")
+
+        @torch.jit.script
+        def foo_decl_always_throws():
+            # type: () -> Tensor
+            raise Exception("Hi")
+
+        output_type = next(foo_decl_always_throws.graph.outputs()).type()
+        self.assertTrue(str(output_type) == "Tensor")
+
+        def foo():
+            raise 3 + 4
+
+        with self.assertRaisesRegex(RuntimeError, "must derive from BaseException"):
+            torch.jit.script(foo)
+
+        # a escapes scope
+        @torch.jit.script
+        def foo():
+            if 1 == 1:
+                a = 1
+            else:
+                if 1 == 1:
+                    raise Exception("Hi")
+                else:
+                    raise Exception("Hi")
+            return a
+        self.assertEqual(foo(), 1)
+
+        @torch.jit.script
+        def tuple_fn():
+            raise RuntimeError("hello", "goodbye")
+
+        with self.assertRaisesRegex(torch.jit.Error, "hello, goodbye"):
+            tuple_fn()
+
+        @torch.jit.script
+        def no_message():
+            raise RuntimeError
+
+        with self.assertRaisesRegex(torch.jit.Error, "RuntimeError"):
+            no_message()
+
+    def test_python_op_exception(self):
+        @torch.jit.ignore
+        def python_op(x):
+            raise Exception("bad!")
+
+        @torch.jit.script
+        def fn(x):
+            return python_op(x)
+
+        with self.assertRaisesRegex(RuntimeError, "operation failed in the TorchScript interpreter"):
+            fn(torch.tensor(4))
+
+    def test_dict_expansion_raises_error(self):
+        def fn(self):
+            d = {"foo": 1, "bar": 2, "baz": 3}
+            return {**d}
+
+        with self.assertRaisesRegex(torch.jit.frontend.NotSupportedError,
+                                    "Dict expansion "):
+            torch.jit.script(fn)
+
+    def test_custom_python_exception(self):
+        class MyValueError(ValueError):
+            def __init__(self, msg):
+                super(MyValueError, self).__init__(msg)
+
+        @torch.jit.script
+        def fn():
+            raise MyValueError("test custom exception")
+
+        with self.assertRaisesRegex(torch.jit.Error, "__torch__.jit.test_exception.MyValueError: test custom exception"):
```

Comment:
Can we test a case where exception is defined outside of `main`? TorchScript handles classes in `main` in a special case by adding a `__torch__` prefix. This is not common for users. It is more useful to guarantee that we can maintain the correct original fully qualified name for exceptions defined in non-main modules. could we add a test for that?

Reply:
The exception is defined in jit.test_exception module actually. Do you actually want me create an exception in a different module to the one raising the exception?

Reply:
Yeah, let's try defining it in a different module and raise it to see if qualified name is easily recognizable.

Reply:
Just tried. The __torch__ prefix is added when the exception module is added elsewhere as well. Looks like here is the code adding the prefix: https://github.com/pytorch/pytorch/blob/d100d98db8424d6d88c9d72d156c49ba94e9d18e/torch/_jit_internal.py#L1029-L1035 .

Will send a update with the extra unit tests after I resolve other comments.

Reply:
I see. This should be fine, as long as we are not losing original module_name. For `main`, the original module name is lost.